packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+                        } = &dep.source
+                        {
+                            let new_component_state = ComponentStateDescription {
+                                component_idx: *component_idx,
+                                state_var_idx: *state_var_idx,
+                            };
+
+                            let new_current_freshness = dep.value.get_freshness();
+
+                            match new_current_freshness {
+                                // No need to recurse if the state var of the dependency is already fresh
+                                // so don't do anything to just continue the inner loop
+                                Freshness::Fresh => (),
+                                Freshness::Stale => {
+                                    // Recurse by putting two items on the stack
+                                    // and then continuing at start of stack loop
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state,
+                                            instruction_idx,
+                                            val_idx,
+                                        },
+                                    ));
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state: new_component_state,
+                                            instruction_idx: 0,
+                                            val_idx: 0,
+                                        },
+                                    ));
+
+                                    // break out of the inner loop and continue to outer loop,
+                                    // which will result in first popping off the last instruction
+                                    // that we added to the stack
+                                    continue 'stack_loop;
+                                }
+                                Freshness::Unresolved => {
+                                    panic!("How did a stale state variable depend on an unresolved state variable?")
+                                }
+                            };
+                        }
+                    }
+
+                    initial_val_idx = 0;
+                }
+
+                // If we have gotten to this part,
+                // then we have freshened all dependencies
+                // for the state variable (component_idx, state_var_idx).
+                // We calculate its value if a dependency has changed,
+                // else restore its previous value.
+
+                let component_idx = component_state.component_idx;
+                let state_var_idx = component_state.state_var_idx;
```

Comment:
You're saving these vars to use exactly once?

Reply:
Once these state variables have been freshened, they are cached and can be used multiple times. This code will need to be run again for a state variable only if one of its dependencies has changed, marking it stale.

The intent of this algorithm is to calculate the state variables (and dependencies) lazily, and once they are need, to memoize the state variable values until they change.

Reply:
No, what I mean is you're aliasing `component_state.component_idx` but you only use the alias one time.

Reply:
I suppose so. This code is now outdated, but I do that elsewhere. I assume it isn't a problem since I imagine the compiler will optimize that out.